rust/connlib/snownet/tests/lib.rs

```diff
@@ -655,48 +684,79 @@ impl TestNode {
+            | Event::ConnectionFailed(_) => None,
+        })
+    }
+
+    fn invalidated_candidates(&self) -> impl Iterator<Item = (u64, Candidate, Instant)> + '_ {
+        self.events.iter().filter_map(|(e, instant)| match e {
+            Event::InvalidateIceCandidate {
                 connection,
                 candidate,
             } => Some((
                 *connection,
                 Candidate::from_sdp_string(candidate).unwrap(),
                 *instant,
             )),
-            Event::ConnectionEstablished(_) | Event::ConnectionFailed(_) => None,
+            Event::NewIceCandidate { .. }
+            | Event::ConnectionEstablished(_)
+            | Event::ConnectionFailed(_) => None,
         })
     }
 
+    fn receive_errors(&self) -> impl Iterator<Item = &snownet::Error> {
+        self.receive_errors.iter()
+    }
+
     fn failed_connections(&self) -> impl Iterator<Item = (u64, Instant)> + '_ {
         self.events.iter().filter_map(|(e, instant)| match e {
             Event::ConnectionFailed(id) => Some((*id, *instant)),
-            Event::SignalIceCandidate { .. } => None,
+            Event::NewIceCandidate { .. } => None,
+            Event::InvalidateIceCandidate { .. } => None,
             Event::ConnectionEstablished(_) => None,
         })
     }
 
     fn receive(&mut self, local: SocketAddr, from: SocketAddr, packet: &[u8], now: Instant) {
-        if let Some((_, packet)) = self
-            .span
-            .in_scope(|| {
-                self.node
-                    .decapsulate(local, from, packet, now, self.buffer.as_mut())
-            })
-            .unwrap()
-        {
-            self.received_packets.push(packet.to_owned())
-        }
+        self.span.in_scope(|| {
+            match self
+                .node
+                .decapsulate(local, from, packet, now, self.buffer.as_mut())
+            {
+                Ok(Some((_, packet))) => self.received_packets.push(packet.to_owned()),
+                Ok(None) => {}
+                Err(e) => self.receive_errors.push(e),
+            }
+        })
     }
 
     fn drain_events(&mut self, other: &mut TestNode, now: Instant) {
         while let Some(v) = self.span.in_scope(|| self.node.poll_event()) {
             self.events.push((v.clone(), now));
 
             match v {
-                Event::SignalIceCandidate {
+                Event::NewIceCandidate {
                     connection,
                     candidate,
-                } => other.node.add_remote_candidate(connection, candidate, now),
+                } => other
+                    .span
+                    .in_scope(|| other.node.add_remote_candidate(connection, candidate, now)),
+                Event::InvalidateIceCandidate {
+                    connection,
+                    candidate,
+                } => other
+                    .span
+                    .in_scope(|| other.node.remove_remote_candidate(connection, candidate)),
```

Comment:
We simulate the control-plane here and thus basically test the entire flow with the assumption that the control plane will deliver this message.